### DISCUSSION TITLE
Eglot doesn't show function signatures and other type information with clangd

### DISCUSSION THREAD

**Discussion by MintSoup** [joaotavora/eglot#1141] [Q&A]
Eglot does not show type information for function signatures in C/C++
when using clangd. It does however work with ccls. lsp-mode does
correctly show type information with clangd though.

I have extracted some segments from the communication between the LSP
server and emacs to help with debugging, hope it will help.

Eglot, CCLS:
```
(:label "push(const value_type &x) -> void" :kind 2 :detail "std::stack" :sortText "   \"" :filterText "push" :insertTextFormat 1 :textEdit
								  (:range
								   (:start
									(:line 7 :character 3)
									:end
									(:line 7 :character 3))
								   :newText "push"))
```

Eglot, clangd:
```
(:detail "[2 overloads]" :filterText "push" :insertText "push" :insertTextFormat 1 :kind 2 :label " push(…)" :score 0.9368076324462891 :sortText "40902d60push" :textEdit
							  (:newText "push" :range
										(:end
										 (:character 3 :line 7)
										 :start
										 (:character 3 :line 7))))
```

lsp-mode, clangd:
```
{
      "detail": "void",
      "filterText": "push",
      "insertText": "push",
      "insertTextFormat": 1,
      "kind": 2,
      "label": "push(value_type &&x)",
      "score": 0.9368076324462891,
      "sortText": "40902d60push",
      "textEdit": {
        "newText": "push",
        "range": {
          "end": {
            "character": 3,
            "line": 7
          },
          "start": {
            "character": 3,
            "line": 7
          }
        }
      }
    }
```

Notice how with lsp-mode/clangd and eglot/ccls, the "label" field is
always filled with the function name along with the correct type
signature, but with eglot/clangd it's just "push(...)".

I am always starting clangd with no arguments on both lsp-mode and eglot.

I'm also attaching some screenshots to show what the completion menus look like.
Eglot:
![eglot](https://user-images.githubusercontent.com/15715189/210221394-0471e9b1-0d9e-49ef-861e-90553aff01dd.png)
lsp-mode:
![lsp](https://user-images.githubusercontent.com/15715189/210221413-b4ae4868-d504-44c5-ac04-8b04fa775198.png)

Is there any way to get eglot to show type information with clangd?

**Comment by jimenezrick** [comment]
Maybe this is also related to this other I reported?

https://github.com/joaotavora/eglot/discussions/1174

**Reply by joaotavora** [reply]
Looks like it, yes.

**Comment by joaotavora** [comment]
@MintSoup I think you're comparing apples to oranges slightly.  I.e. the completions items you're looking at are not _exactly_ the same. Can you do the same comparison and extraction of events lsp-mode/clangd vs eglot/clangd for the `swap` completion? That would help me figure out what is happening.

**Reply by MintSoup** [reply]
eglot/clangd:

```
(:detail "void" :filterText "swap" :insertText "swap" :insertTextFormat 1 :kind 2 :label " swap(stack<int> &s)" :score 0.9368076324462891 :sortText "40902d60swap" :textEdit
							  (:newText "swap" :range
										(:end
										 (:character 3 :line 5)
										 :start
										 (:character 3 :line 5))))
```

lsp-mode/clangd:

```
    {
      "detail": "void",
      "filterText": "swap",
      "insertText": "swap",
      "insertTextFormat": 1,
      "kind": 2,
      "label": "swap(stack<int> &s)",
      "score": 0.9368076324462891,
      "sortText": "40902d60swap",
      "textEdit": {
        "newText": "swap",
        "range": {
          "end": {
            "character": 3,
            "line": 5
          },
          "start": {
            "character": 3,
            "line": 5
          }
        }
      }
    },
```
Interestingly enough `label` is the same for both, but eglot shows `swap [2 overloads]` while lsp-mode shows `swap(stack<int> &s) void (Method)`. Maybe eglot is passing `insertText` or `filterText` to the completion frontend?

**Reply by MintSoup** [reply]
Also I tried using clangd with `--completion-style=detailed` and eglot didn't show me extra completion options for the overloads, while lsp-mode did.

**Reply by jvillasante** [reply]
I'm having the same issue when enabling `--completion-style=detailed` with clangd and eglot won't show overloads, it will only show the method name and "2 overloads" as seen in the screenshot below. `lsp-mode` is able to show all the overloads with the full method definition.

![Screenshot from 2023-03-21 13-58-53](https://user-images.githubusercontent.com/1139125/226904242-1effcd07-97df-43cc-8322-b22a91d38b12.png)

For context, here's `lsp-mode` showing the overloads:

![Screenshot from 2023-03-22 08-27-40](https://user-images.githubusercontent.com/1139125/226904916-79a8818e-7c91-4649-8e7f-0cf12d381422.png)

**Reply by joaotavora** [reply]
I can't reproduce this:

```
clangd version 15.0.7
Features: linux
Platform: x86_64-pc-linux-gnu
```

```
/path/to/emacs-29/src/emacs -Q -f package-initialize -f global-company-mode ~/tmp/thingy.cpp
```

where `thingy.cpp` is:
```c++
#include <vector>
int main () {
  std::vector<int> vi;
  vi.
}
```

```
C-u M-x eglot RET clangd --completion-style=detailed RET
``` 

type `pu` and get this

![image](https://user-images.githubusercontent.com/387011/226914866-b3361866-04d5-481d-9fd5-800335620bce.png)

**Reply by jvillasante** [reply]
Ok, found my issue. Not sure why I got `eglot` `clangd` configured like this:
```
(add-to-list 'eglot-server-programs
        '(c-mode c++-mode
             . ("clangd"
                   "-j=8"
                   "--log=error"
                   "--malloc-trim"
                   "--background-index"
                   "--clang-tidy"
                   "--cross-file-rename"
                   "--completion-style=detailed"
                   "--pch-storage=memory"
                   "--header-insertion=never"
                   "--header-insertion-decorators=0")))
```

The issue goes away when I only have one mode in there: 
```
(add-to-list 'eglot-server-programs
        '(c++-mode
             . ("clangd"
                   "-j=8"
                   "--log=error"
                   "--malloc-trim"
                   "--background-index"
                   "--clang-tidy"
                   "--cross-file-rename"
                   "--completion-style=detailed"
                   "--pch-storage=memory"
                   "--header-insertion=never"
                   "--header-insertion-decorators=0")))
```

Thanks!

**Comment by joaotavora** [comment]
I've reproduced and pushed a fix to this problem in Emacs 29's `lisp/progmodes/eglot.el`. Check it out and report back here (or in a new Emacs bug report) if it works.

**Reply by joaotavora** [reply]
See https://git.savannah.gnu.org/cgit/emacs.git/commit/?h=emacs-29&id=e3be0dbf85c729447776d361ba56ada6b92f0149

**Reply by MintSoup** [reply]
Thank you, I think it's almost working perfectly. Just one thing:
![image](https://user-images.githubusercontent.com/15715189/220992021-ca5443b2-0615-4afe-845e-3be53ae49122.png)
Notice the weird dots in front of some of the completions? It also happens with `emacs -q`:
![image](https://user-images.githubusercontent.com/15715189/220992456-2115c934-9fd4-4ec5-b30b-30161dee2546.png)
Interestingly, when I type `std::` in a C++ buffer (no includes), all completion candidates have the dot in the front. Can you reproduce this?

**Reply by joaotavora** [reply]
> Can you reproduce this?

Yes, I can :-) It confused me at first too, but now I think I understand what it means.  It means that i you select that completion, the LSP server has also provided additional includes to be automatically added for you.  Completions without the dot will only change the local part.

I find this modestly useful, I must say.

**Reply by MintSoup** [reply]
Ah, clangd even has it documented: it's the `--header-insertion-decorators` which can be enabled/disabled.
